### DISCUSSION TITLE
alos2App ionospheric correction errors

### DISCUSSION THREAD

**Discussion by EJFielding** [isce-framework/isce2#671] [Q&A] [answered]
I am processing two frames of ALOS-2 ScanSAR data (WD1, 5-subswaths). One of the subswaths of one of the frames is showing up with a different ionospheric correction estimate. I tried changing the number of looks in the ionospheric estimation, but I get the same result. I attach an image of the estimate before filtering (ion_20rlks_224alks.ion). I don't see this discontinuity in either the lower or upper interferogram phase, so I don't know where it is coming from, but it causes errors in the final ionospheric correction near the area where the discontinuity is located.

Is this coming from the original stitching of the subswaths and frames? I notice that there is a data window position shift between the two frames.
@CunrenLiang  
![ion_20rlks_224alks-ion](https://user-images.githubusercontent.com/7253576/221339281-516c45cd-d568-4893-a36e-5a6dcf981048.jpg)

**Comment by CunrenLiang** [comment, marked as answer]
The swath phase differences now use some default values stored in the scripts. These are values estimated using a large number of datasets, so they are more precise. However, occasionally the actual values are different from these default values. Subswath five of the second frame in this case apparently has a different phase difference, so we have to use the estimated value. See below for details of settings.
https://github.com/isce-framework/isce2/blob/main/examples/input_files/alos2/alos2App.xml#L311-L317

after that we need to rerun from ion_subband.

**Reply by EJFielding** [reply]
Thanks! I added this:
`    <property name="swath phase difference snap to fixed values">[[True, True, True, True], [True, True, True, False]]</property>`
Now it looks fixed.
![ion_20rlks_224alks-ion-fixed](https://user-images.githubusercontent.com/7253576/221375701-95d8a5f9-ef35-43ee-8d42-871e63f49227.png)

**Comment by fdelgadodelapuente** [comment]
Hi Eric and Cunren,

I need to apply this option to correct a phase jump in the ion_80rlks_448alks.ion file between swaths 3 and 4

![ion_80rlks_448alks ion](https://github.com/isce-framework/isce2/assets/23100180/78e2abdf-99f6-4087-a7b2-366e1b3b7e4d)

When I set the relevant options to

    <property name="swath phase difference snap to fixed values">[True, True, False, True]</property>

I get the error
  File "/home/fdelgado/isce/isce2-2.6.3/install/isce/components/isceobj/Alos2Proc/runIonSubband.py", line 373, in runIonSubband
    raise Exception('please specify each frame for parameter: swath phase difference snap to fixed values')
Exception: please specify each frame for parameter: swath phase difference snap to fixed values

The error seems straightforward to correct, but the software keeps crashing. How can I fix it?

Thanks a lot

**Reply by EJFielding** [reply]
Hi Francisco, How many frames are you processing? I was using two frames and five subswaths, so I had to specify 8 swath overlaps in a 4 by 2 array. It might be necessary to put in the extra brackets for a 4 by 1 array if you are processing one frame.

**Comment by fdelgadodelapuente** [comment]
Hi Eric,

It is just one frame. I followed your advice by setting the flag to

    <property name="swath phase difference snap to fixed values">[[True, True, False, True], []]</property>

because the jump is between swaths 3 and 4.  But then I obtained the same error than before. Then I switched it to 

    <property name="swath phase difference snap to fixed values">[[True, True, False, True]]</property>

And that solved the problem.

![Screen Shot 2023-10-09 at 11 17 54](https://github.com/isce-framework/isce2/assets/23100180/de979486-2fef-49df-8fa4-5d27c15fdf55)

I don't know why but the double brackets made the difference. 

Thanks for your help

**Reply by EJFielding** [reply]
Yes, my description was not precise. The second version you used is what I meant. I think it is expecting a 2-dimensional array, so you need the two brackets even if the "frame" dimension is equal to 1.
